runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -2562,11 +2587,8 @@ J9::Z::TreeEvaluator::checkcastEvaluator(TR::Node * node, TR::CodeGenerator * cg
                if (comp->getOption(TR_TraceCG))
                   traceMsg(comp, "%s: Emitting NullTest\n", node->getOpCode().getName());
                TR_ASSERT(!objectNode->isNonNull(), "Object is known to be non-null, no need for a null test");
-               bool isNullTestImplicit = genInstanceOfOrCheckCastNullTest(node, cg, objectReg);
-               if (!isNullTestImplicit)
-                  {
-                  generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BE, node, doneLabel);
-                  }
+               genInstanceOfOrCheckCastNullTest(node, cg, objectReg);
+               generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BE, node, doneLabel);
```

Comment:
We do not need to generate a Branch here if node is `checkcastAndNULLCHK`, as we are handling this case in `genInstanceOfOrCheckCastNullTest`

Reply:
Right, totally forgot. I'm going to open an issue for this. What ends up happening is we generate something like this:

```
------------------------------
 n223n    (  0)  checkcastAndNULLCHK [#84]                                                            [0x5022493570] bci=[-1,78,-] rc=0 vc=556 vn=- li=21 udi=- nc=2
 n388n    (  1)    ==>aRegLoad (in &GPR_0448) (SeenRealReference sharedMemory )
 n129n    (  0)    loadaddr  java/lang/Integer[#361  Static] [flags 0x18307 0x0 ] (in GPR_0451)       [0x5022491810] bci=[-1,78,-] rc=0 vc=556 vn=- li=21 udi=- nc=0
------------------------------
   [0x50226B97A0] LGFI    GPR_0451,687681792
 [0x50226B98C0] LTGR    &GPR_0448,&GPR_0448
 [0x50226B9A10] BRC     BH(0x8), Snippet Label L0224
 [0x50226B9B80] BRC     BH(0x8), Label L0208
 [0x50226B9DB0] LLZRGF  GPR_0464,#413 0(&GPR_0448)
 [0x50226B9EA0] CGRJ    GPR_0451,GPR_0464,Label L0208,BH(mask=0x8),
 [0x50226BA160] BRC     NOP(0xf), Outlined Label L0209
 [0x50226BD8E0] ASSOCREGS
 [0x50226BD310] Label L0208:
 POST:
 {AssignAny:&GPR_0448:R} {AssignAny:GPR_0464:R} {AssignAny:GPR_0451:R}
  ------------------------------
```

The two `BRC` are redundant as you mention. This is a general pattern which should be caught in a peephole. I'll ask Simona to work on this one.

Reply:
We can get rid of branch here without living it to peephole. 
```
if (node->getOpCodeValue() != TR::checkcastAndNULLCHK)
   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BE, node, doneLabel);
```

Reply:
Fixed in ff790c2 to avoid generating the double branch.